### ISSUE TITLE
Fix FP S3827 (`no-reference-error`): activate for TS and improve

### ISSUE THREAD

**Issue by ilia-kebets-sonarsource** [SonarSource/SonarJS#4514]
This rule is not activated for TS because we expected these kinds of issues to be caught by the TS compiler.
However, there are cases where they are not, such as this one:

```ts
const arr = [1,2,3,4];
const res = arr.map(item => invalidData.includes(item));
const invalidData = [1,2];
```
We should catch these cases if possible.

This issue was raised in [this community ticket](https://community.sonarsource.com/t/s3827-variables-should-be-defined-before-being-used-for-typescript/106569).

**Comment by Wohops** [comment]
Migrated to Jira. New ticket number: JS-39
